flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunner.java

```diff
@@ -261,34 +261,69 @@ public void grantLeadership(UUID leaderSessionID) {
 
     @GuardedBy("lock")
     private void startJobMasterServiceProcessAsync(UUID leaderSessionId) {
-        sequentialOperation =
-                sequentialOperation.thenRun(
-                        () ->
-                                runIfValidLeader(
-                                        leaderSessionId,
-                                        ThrowingRunnable.unchecked(
-                                                () ->
-                                                        verifyJobSchedulingStatusAndCreateJobMasterServiceProcess(
-                                                                leaderSessionId)),
-                                        "verify job scheduling status and create JobMasterServiceProcess"));
-
+        boolean isValid;
+        synchronized (lock) {
+            isValid = isValidLeader(leaderSessionId);
+        }
+        if (isValid) {
+            sequentialOperation =
+                    sequentialOperation.thenCompose(
+                            unused ->
+                                    runIfValidLeader(
+                                                    leaderSessionId,
+                                                    () ->
+                                                            jobResultStore.hasJobResultEntryAsync(
+                                                                    getJobID()),
+                                                    "verify jbb result entry")
+                                            .handle(
+                                                    (hasJobResult, throwable) -> {
+                                                        if (hasJobResult == null) {
```

Comment:
This would swallow a potential IOException in the `hasJobResultEntryAsync` call. The old implementation handled these kind of errors. PTAL

Reply:
Currently the `throwable` is handled.

Reply:
The `if (hasJobResult == null) {` block becomes obsolete here (because there shouldn't be a situation where no result is passed if no error happened in upstream async calls for `handle(BiFunction)`. You could add a `Precondition.checkNotNull` if you like. But the subsequent if call would cause a `NullPointerException` anyway which makes the precondition obsolete again.

(... and a side-remark here: we would at least add logs rather than only returning `null` in case if the `if (hasJobResult == null) {` case.)

Reply:
~The if (hasJobResult == null) { block becomes obsolete here (because there shouldn't be a situation where no result is passed if no error happened in upstream async calls for handle(BiFunction). You could add a Precondition.checkNotNull if you like. But the subsequent if call would cause a NullPointerException anyway which makes the precondition obsolete again.~

~(... and a side-remark here: we would at least add logs rather than only returning null in case if the if (hasJobResult == null) { case.)~

Ok, I misread the code here. The `null` check actually has a meaning here because the upstream call of `supplyAsyncIfValidLeader` call returns a `CompletableFuture` that completed successfully with `null` if the leadership is not granted anymore (the `noLeaderFallback`). This functional dependency is a bit hidden (IMHO). I see two options to improve the code:
1. Add a comment to the `if (hasJobResult == null) {` explaining why this is needed. I would also like to add a `@Nullable` annotation to `printLogIfNotValidLeader` to acknowlegde. But I don't know how to handle such a case with async calls. The `@Nullable` annotation becomes kind of ambigious.
2. Consider not having the leadership in the `supplyAsyncIfValidLeader` an error scenario that shall be handled in the downstream `handle` call. This we can achieve by returning an exceptionally completed future. You might want to introduce a new exception `LeadershipLostException` that derives from `LeaderElectionException` to cover that error scenario. You could add the log message in the `handle` method when handling `throwable != null`, something like:
```
if (throwable instanceof LeadershipLostException) {
    // print your log statement
} else if (throwable != null) {
    ExceptionUtils.rethrow(throwable);
}
```

I'm in favor of the 2nd solution because it makes the code more explicit and, in this regard, easier to understand.